File: core/brir_layout.py
"""Fixed BRIR layouts with optional, silent extension pairs removed.

HeSuVi always consumes fourteen positional IR channels. Equalizer APO cycles
shorter IRs, so even a stereo-only measurement needs those fourteen slots.
The generic HRIR layout keeps its sixteen base slots, including LFE.
"""

from __future__ import annotations

import json
import struct
from collections.abc import Sequence
from pathlib import Path

import numpy as np

from core.constants import HESUVI_TRACK_ORDER, HEXADECAGONAL_TRACK_ORDER


def base_channel_count(order: Sequence[str]) -> int:
    """Return the fixed base size; custom output layouts are kept in full."""
    if tuple(order) == tuple(HESUVI_TRACK_ORDER):
        return 14
    if tuple(order) == tuple(HEXADECAGONAL_TRACK_ORDER):
        return 16
    return len(order)


def trim_silent_extensions(data: np.ndarray, order: Sequence[str]) -> np.ndarray:
    """Trim only trailing zero extension pairs, never holes or quiet responses."""
    minimum = base_channel_count(order)
    end = len(order)
    while end > minimum and not np.any(data[end - 2:end] != 0):
        end -= 2
    return data[:end]


def compact_tracks(data: np.ndarray, order: Sequence[str]) -> tuple[np.ndarray, tuple[str, ...]]:
    """Opt-in removal of all zero rows, including positional placeholders."""
    keep = np.any(data != 0, axis=1)
    if not np.any(keep):
        raise ValueError("All BRIR channels are silent; no channels remain to write.")
    return data[keep], tuple(name for name, active in zip(order, keep) if active)


def append_track_names(path, names: Sequence[str]) -> None:
    """Embed a versioned ICHL JSON channel map in a freshly written RIFF WAV."""
    payload = json.dumps({"version": 1, "tracks": list(names)}, separators=(",", ":")).encode()
    chunk = b"ICHL" + struct.pack("<I", len(payload)) + payload
    chunk += b"\0" * (len(payload) % 2)
    with Path(path).open("r+b") as handle:
        header = handle.read(12)
        if header[:4] != b"RIFF" or header[8:12] != b"WAVE":
            raise ValueError("Compact BRIR metadata requires a RIFF WAV file.")
        size = handle.seek(0, 2)
        if size + len(chunk) - 8 > 0xFFFFFFFF:
            raise ValueError("Compact BRIR exceeds the RIFF WAV size limit.")
        handle.write(chunk)
        handle.seek(4)
        handle.write(struct.pack("<I", size + len(chunk) - 8))


def read_track_names(path, canonical_order: Sequence[str], channel_count: int) -> tuple[str, ...] | None:
    """Validate an embedded map; None means a legacy positional layout.

    Map validation precedes positional inference, including compact files that
    happen to have 14/16 channels. Unknown RIFF chunks are skipped by seeking.
    """
    found = None
    with Path(path).open("rb") as handle:
        header = handle.read(12)
        if header[:4] != b"RIFF" or header[8:12] != b"WAVE":
            return None
        end = min(struct.unpack("<I", header[4:8])[0] + 8, Path(path).stat().st_size)
        while handle.tell() + 8 <= end:
            chunk, size = struct.unpack("<4sI", handle.read(8))
            next_chunk = handle.tell() + size + size % 2
            if next_chunk > end:
                raise ValueError("Truncated WAV chunk.")
            if chunk == b"ICHL":
                if found is not None or size > 4096:
                    raise ValueError("Duplicate or oversized BRIR channel mapping.")
                try:
                    value = json.loads(handle.read(size))
                except (ValueError, UnicodeError) as exc:
                    raise ValueError("Invalid BRIR channel mapping JSON.") from exc
                if not isinstance(value, dict) or type(value.get("version")) is not int or value["version"] != 1:
                    raise ValueError("Unsupported BRIR channel mapping version.")
                names = value.get("tracks")
                if (
                    not isinstance(names, list)
                    or not all(isinstance(name, str) for name in names)
                    or len(names) != channel_count
                    or len(set(names)) != len(names)
                    or any(name not in canonical_order for name in names)
                ):
                    raise ValueError("BRIR channel mapping does not match the WAV layout.")
                found = tuple(names)
            handle.seek(next_chunk)
    return found

File: crates/impulcifer-io/src/brir_layout.rs
//! Versioned compact BRIR channel maps, core/brir_layout.py:46-100.
use crate::IoError;
use std::fs::{File, OpenOptions};
use std::io::{Read, Seek, SeekFrom, Write};
use std::path::Path;

fn invalid(message: &str) -> IoError {
    IoError::Format(message.into())
}

pub fn append_track_names(path: &Path, names: &[&str]) -> Result<(), IoError> {
    // Explicit field order matches Python's insertion-ordered JSON object.
    let names = serde_json::to_string(names).map_err(|e| invalid(&e.to_string()))?;
    let payload = format!("{{\"version\":1,\"tracks\":{names}}}").into_bytes();
    let mut file = OpenOptions::new().read(true).write(true).open(path)?;
    let mut header = [0; 12];
    file.read_exact(&mut header)?;
    if &header[..4] != b"RIFF" || &header[8..] != b"WAVE" {
        return Err(invalid("Compact BRIR metadata requires a RIFF WAV file."));
    }
    let end = file.seek(SeekFrom::End(0))?;
    let size = u32::try_from(end + payload.len() as u64 + (payload.len() % 2) as u64)
        .map_err(|_| invalid("Compact BRIR exceeds the RIFF WAV size limit."))?;
    file.write_all(b"ICHL")?;
    file.write_all(&(payload.len() as u32).to_le_bytes())?;
    file.write_all(&payload)?;
    if payload.len() % 2 != 0 {
        file.write_all(&[0])?;
    }
    file.seek(SeekFrom::Start(4))?;
    file.write_all(&size.to_le_bytes())?;
    Ok(())
}

pub fn read_track_names(
    path: &Path,
    canonical_order: &[&str],
    channel_count: usize,
) -> Result<Option<Vec<String>>, IoError> {
    let mut file = File::open(path)?;
    let physical = file.metadata()?.len();
    if physical < 12 {
        return Ok(None);
    }
    let mut header = [0; 12];
    file.read_exact(&mut header)?;
    if &header[..4] != b"RIFF" || &header[8..] != b"WAVE" {
        return Ok(None);
    }
    let end = (u64::from(u32::from_le_bytes(header[4..8].try_into().unwrap())) + 8).min(physical);
    let mut found = None;
    while file.stream_position()? + 8 <= end {
        let mut chunk = [0; 8];
        file.read_exact(&mut chunk)?;
        let size = u32::from_le_bytes(chunk[4..].try_into().unwrap()) as u64;
        let next = file.stream_position()? + size + size % 2;
        if next > end {
            return Err(invalid("Truncated WAV chunk."));
        }
        if &chunk[..4] == b"ICHL" {
            if found.is_some() || size > 4096 {
                return Err(invalid("Duplicate or oversized BRIR channel mapping."));
            }
            let mut bytes = vec![0; size as usize];
            file.read_exact(&mut bytes)?;
            let value: serde_json::Value = serde_json::from_slice(&bytes)
                .map_err(|_| invalid("Invalid BRIR channel mapping JSON."))?;
            if value["version"].as_u64() != Some(1) {
                return Err(invalid("Unsupported BRIR channel mapping version."));
            }
            let names = value["tracks"]
                .as_array()
                .and_then(|a| {
                    a.iter()
                        .map(|v| v.as_str().map(str::to_owned))
                        .collect::<Option<Vec<_>>>()
                })
                .ok_or_else(|| invalid("BRIR channel mapping does not match the WAV layout."))?;
            let mut seen = std::collections::HashSet::new();
            if names.len() != channel_count
                || names
                    .iter()
                    .any(|n| !canonical_order.contains(&n.as_str()) || !seen.insert(n))
            {
                return Err(invalid(
                    "BRIR channel mapping does not match the WAV layout.",
                ));
            }
            found = Some(names);
        }
        file.seek(SeekFrom::Start(next))?;
    }
    Ok(found)
}
